src/main/java/uk/gov/ons/ctp/response/action/scheduled/distribution/ActionDistributor.java

```diff
@@ -22,101 +22,105 @@
+      @Qualifier("social") ActionProcessingService socialActionProcessingService) {
+    this.appConfig = appConfig;
+    this.redissonClient = redissonClient;
+    this.actionRepo = actionRepo;
+    this.actionCaseRepo = actionCaseRepo;
+    this.actionTypeRepo = actionTypeRepo;
+    this.caseSvcClientService = caseSvcClientService;
+    this.businessActionProcessingService = businessActionProcessingService;
+    this.socialActionProcessingService = socialActionProcessingService;
+  }
 
   /**
-   * wake up on schedule and check for submitted actions, enrich and distribute them to spring
-   * integration channels
-   *
-   * @return the info for the health endpoint regarding the distribution just performed
+   * Called on schedule to check for submitted actions then creates and distributes requests to
+   * action exporter or notify gateway
    */
   @Transactional
-  public DistributionInfo distribute() {
-    log.debug("ActionDistributor awoken...");
-    final DistributionInfo distInfo = new DistributionInfo();
-    final List<ActionType> actionTypes = actionTypeRepo.findAll();
-
-    for (final ActionType actionType : actionTypes) {
-      final List<InstructionCount> instructionCounts = processActionType(actionType);
-      distInfo.getInstructionCounts().addAll(instructionCounts);
-    }
-
-    log.debug("ActionDistributor going back to sleep");
-    return distInfo;
+  public void distribute() {
+    List<ActionType> actionTypes = actionTypeRepo.findAll();
+    actionTypes.forEach(this::processActionType);
   }
 
-  private List<InstructionCount> processActionType(final ActionType actionType) {
-    log.with("action_type", actionType.getName()).debug("Dealing with actionType");
-    final InstructionCount requestCount =
-        InstructionCount.builder()
-            .actionTypeName(actionType.getName())
-            .instruction(DistributionInfo.Instruction.REQUEST)
-            .count(0)
-            .build();
-    final InstructionCount cancelCount =
-        InstructionCount.builder()
-            .actionTypeName(actionType.getName())
-            .instruction(DistributionInfo.Instruction.CANCEL_REQUEST)
-            .count(0)
-            .build();
-
+  private void processActionType(final ActionType actionType) {
+    String actionTypeName = actionType.getName();
+    RLock lock = redissonClient.getFairLock(lockPrefix + actionTypeName);
     try {
-      final List<Action> actions = retrieveActions(actionType);
-      processActions(actions, requestCount, cancelCount);
-    } catch (final Exception e) {
-      log.with("action_type", actionType.getName()).error("Failed to process action type", e);
+      // Wait for a lock. Automatically unlock after a certain amount of time to prevent issues
+      // when lock holder crashes or Redis crashes causing permanent lockout
+      if (lock.tryLock(
+          appConfig.getDataGrid().getLockTimeToWaitSeconds(),
```

Comment:
in actual fact - and I got this wrong in Sample - it might be preferable to skip the wait, because if one type is locked then we can just work on another one, without waiting for the other process to relinquish the lock. It'll be more performant

Reply:
Makes sense I'll remove the wait